architecture/spmc.rst

```diff
@@ -0,0 +1,267 @@
+                deactivate spmc_sp_handler
+                spmc_sp_handler -> spmc_sp_handler: Retrieve FF-A message from SP context
+                return
+                return
+                return
+        end
+                return
+        == End of thread ==
+                return
+        return \n <font color=red>SMC
+        return FFA_MSG_SEND_DIRECT_RESP \n <font color=red>ERET
+
+
+Every message received by the SPMC from the Normal World is handled in the
+``thread_spmc_msg_recv()`` function.
+
+When entering a SPs we need to be running on a thread. This is needed to be
+able to push the TS session (We push the TS session to get access to the SP
+memory).
+Currently the only possibility to enter a SP from the Normal world is via a
+``FFA_MSG_SEND_DIRECT_REQ``. Whenever we receive a ``FFA_MSG_SEND_DIRECT_REQ``
+message which doesn't have OP-TEE as the endpoint-id, we start a thread and
+forward the FF-A message to the ``sp_msg_handler()``.
+
+The ``sp_msg_handler()`` is responsible for all messages coming or going
+to/from a SP. It runs in a while loop and will handle every message until it
+comes across a messages which is not intended for the secure world.
+After a message is handled by the SPMC or when it needs to be forwarded to a SP,
+``sp_enter()`` is called.
+``sp_enter()`` will copy the FF-A arguments and resume the SP.
+
+When the SPMC needs to have access to the SPs memory, it will call
+``ts_push_current_session()`` to gain access and ts_pop_current_session to
+release the access.
+
+Running and exiting SPs
+-----------------------
+
+The SPMC resumes/starts the SP by calling the sp_enter(). This will set-up the
+SP context and jump into S-EL0.
+Whenever the SP performs a system call it will end-up in ``sp_handle_svc()``.
+``sp_handle_svc()`` store the current context of the SP and makes sure that we
+don't return to S-EL0 but instead returns to S-EL1 back to ``sp_enter()``.
+``sp_enter()`` will pass the FF-A registers (x0-x7) to
+``spmc_sp_msg_handler()``. This will process the FF-A message.
+
+
+RxTx buffer managment
+---------------------
+RxTx buffers are used by the SPMC to exchange information between an endpoint
+and the SPMC. The rxtx_buf struct is used by the SPMC for abstracting buffer
+management.
+Every SP has a ``struct rxtx_buf`` wich will be passed to every function that
+needs access to the rxtx buffer.
+A separate ``struct rxtx_buf`` is defined for the Normal World, which gives
+access to the Normal World buffers.
+
+
+Configuration
+=============
+
+Adding SPs to the Image
+-----------------------
+
+The following flags have to be enabled to enable the SPMC. The SP images
+themself are loaded by using the ``SP_PATHS`` flag.
+
+.. code-block:: Make
+
+	OPTEE_OS_COMMON_FLAGS += CFG_WITH_SPMC=y CFG_CORE_FFA=y
+	OPTEE_OS_COMMON_FLAGS += CFG_WITH_SP=y
+	OPTEE_OS_COMMON_FLAGS += SP_PATHS="SP.elf"
+	TF_A_FLAGS += SPD=spmd SPMD_SPM_AT_SEL2=0
```

Comment:
`OPTEE_OS_COMMON_FLAGS` relates to a config from OP-TEE/build.git. Maybe worth mentionning that here: or simply state that OP-TEE configuration should have:
`CFG_WITH_SPMC=y`
`CFG_CORE_FFA=y`
`CFG_WITH_SP=y`
`SP_PATHS="path/to/sp-xxx.elf path/to/sp-yyy.elf`

(by the way, it would be more consistent to rename `SP_PATHS` to `CFG_SP_PATHS` in optee_os.git)

Reply:
I suppose `SP_PATHS` follows `EARLY_TA_PATHS`

Reply:
true :)  both should be `CFG_`s IMO. But this is probably not a real issue.

Reply:
> true :) both should be `CFG_`s IMO. But this is probably not a real issue.

No, it is on purpose that they are not `CFG_*` in order to make it possible to not build the "TA dev kit" twice, see https://github.com/OP-TEE/optee_os/blob/3.12.0/mk/config.mk#L297-L303 (indeed, defining a new `CFG_` would cause recompilation). Probably not very important though.

Reply:
ok, fine. i had forgotten this. i found back your [comment](https://github.com/OP-TEE/optee_os/pull/1733#discussion_r131632045).

Reply:
`OPTEE_OS_COMMON_FLAGS` are still there